idb/idb.go

```diff
@@ -39,29 +45,56 @@ type TxnRow struct {
 }
 
 // Next returns what should be an opaque string to be returned in the next query to resume where a previous limit left off.
-func (tr TxnRow) Next() string {
-	var b [12]byte
-	binary.LittleEndian.PutUint64(b[:8], tr.Round)
-	binary.LittleEndian.PutUint32(b[8:], uint32(tr.Intra))
-	return base64.URLEncoding.EncodeToString(b[:])
+func (tr TxnRow) Next() (string, error) {
+	var b [12 + crypto.DigestSize]byte
+	binary.LittleEndian.PutUint64(b[crypto.DigestSize:40], tr.Round)
+	binary.LittleEndian.PutUint32(b[40:], uint32(tr.Intra))
+
+	// avoid in-place decoding to prevent bounds overwrite. Could optimize with DecodeLen.
+	data, err := base32.StdEncoding.WithPadding(base32.NoPadding).DecodeString(tr.Extra.RootTxid)
+	if err != nil {
+		return "", fmt.Errorf("Next() err: %w", err)
+	}
+	if len(data) != 0 {
+		if len(data) != crypto.DigestSize {
+			return "", fmt.Errorf("Next() unexpected txid size")
+		}
+		copy(b[:crypto.DigestSize], data)
+	}
+
+	return base64.URLEncoding.EncodeToString(b[:]), nil
 }
 
 // DecodeTxnRowNext unpacks opaque string returned from TxnRow.Next()
-func DecodeTxnRowNext(s string) (round uint64, intra uint32, err error) {
-	var b []byte
-	b, err = base64.URLEncoding.DecodeString(s)
+func DecodeTxnRowNext(s string) (round uint64, intra uint32, rootTxid string, err error) {
```

Comment:
If we wanted to be clever we could decode the previous version of next-token here based on the decoded length; but I don't care enough to press it and approximately zero API paging sequences will be disrupted.

Reply:
Ok, it never occurred to me to consider someone saving the next-token string between releases. I think I'm comfortable letting that generate an error.